### ISSUE TITLE
Windows native dialogs reversion

### ISSUE THREAD

**Issue by criptych** [geany/geany#578]
I noticed after upgrading to 1.25 that the Windows native file dialogs have reverted to a 
"Windows 95" look & feel where in 1.24 they had a true "native" L&F (I'm using Windows 7).  Most annoying about this is that the dialogs are tiny (about 400x300) and not resizable.  It's possible this is an issue with GTK but I don't know where to start to investigate that.

In Geany 1.24:
![geany124-filedialog](https://cloud.githubusercontent.com/assets/1134921/8750266/7dab004a-2c74-11e5-9959-3d8f38909baf.png)

In Geany 1.25:
![geany125-filedialog](https://cloud.githubusercontent.com/assets/1134921/8750267/7daf80a2-2c74-11e5-8c92-bc6575a8de05.png)
Note the missing sidebar and resize handle, as well as the older-style toolbar icons and layout.  (On the bright side, it doesn't leave trails on the editor window when I move it. :)

Please note this is just an aesthetic issue; overall I love the new version (the message bar for changed files is nice, and glad to see it no longer comments blank lines as this didn't play nice with "strip trailing spaces"!).

**Comment by codebrainz** [comment]
For anyone else wondering, I think L&F might mean "look and feel".

@eht16 @kugel- @zhekov isn't this controlled by one of those defines before including `windows.h`?

**Comment by criptych** [comment]
Yes, that's what I meant.  Sorry for not clarifying.

**Comment by b4n** [comment]
I have no idea of the cause, but I can confirm this on a Windows 8 machine.

**Comment by zhekov** [comment]
You may wish to skip to the final TL;DR paragraph below.

The look is normally controlled by the OFN_EXPLORER flag in the OPENFILENAME structure passed to GetOpenFileName and GetSaveFileName. However:

"For compatibility reasons, the Places Bar is hidden if Flags is set to OFN_ENABLEHOOK and lStructSize is OPENFILENAME_SIZE_VERSION_400."

To my surprise, we set lStructSize to OPENFILENAME_SIZE_VERSION_400 if the latter is defined. And it is defined in API level 0x0500 and above: the purpose of this constant is to request the open/save dialogs 0x0400 for a program compiled with API 0x0500 or greater. I don't know the exact purpose - maybe to do something NT4/9X specific when the program is started under these systems [1]. Normally, they'll just ignore a larger OPENFILENAME structure, and use only the members they know about. What I'm sure about is that the usage is very rare, and we shouldn't need it.

[1] As clarified in PR #573, simply defining WINVER/_WIN32_WINNT to N does not make your program will require windows N. You can compile with the XP API, but if you are not using any XP functions, it will run on lower Windows versions.

TL;DR

I removed all "#ifdef OPENFILENAME_SIZE_VERSION_400 lStructSize = OPENFILENAME_SIZE_VERSION_400" from win32.c, and the dialogs now look fine. (Though I'm normally using the gtk+ dialogs - probably got used to them.)

**Comment by zhekov** [comment]
One more clarification: we are using OFN_ENABLEHOOK, the hook function is required to update the gtk+ UI while using a modal Win~1 dialog.

**Comment by b4n** [comment]
> "For compatibility reasons, the Places Bar is hidden if Flags is set to OFN_ENABLEHOOK and lStructSize is OPENFILENAME_SIZE_VERSION_400."

Wow okay so when @codebrainz added `OFN_ENABLEHOOK` to fix the UI freezes, it implicitly lead to using the old dialogs while when using `OPENFILENAME_SIZE_VERSION_400` it didn't?  So obvious… :D

Anyway, so if I read you correctly, you say we can just remove the `OPENFILENAME_SIZE_VERSION_400` references and all will be good?  Cant this cause problems if building e.g. on Windows 8 and trying to run on e.g. XP?  (as the size would be the one from Windows 8)

**Comment by eht16** [comment]
I can confirm dropping OPENFILENAME_SIZE_VERSION_400 removes the old dialogs.
http://pastebin.geany.org/84NwN/ this is what I tried.

Ditching in my memories, I think I added these OPENFILENAME_SIZE_VERSION_400 probably because I read somewhere this might be useful or fix something or so. But this is many, many years ago and especially if this was related to pre-WinXP systems, it is just obsolete because we don't support such old versions anymore.

**Comment by criptych** [comment]
> I have no idea of the cause, but I can confirm this on a Windows 8 machine.

At least I'm not going crazy. :)

> (Though I'm normally using the gtk+ dialogs - probably got used to them.)

I don't mind them, but I occasionally find it handy that the native dialogs let you do everything Explorer does (copy, move, rename) without leaving the dialog.

Thanks for looking into this.

**Comment by zhekov** [comment]
> so when @codebrainz added OFN_ENABLEHOOK to fix the UI freezes, it implicitly lead to using the old dialogs while when using OPENFILENAME_SIZE_VERSION_400 it didn't?

You need both. BTW, the dialog is still *not resizable*, seems that we need *specify OFN_ENABLESIZING* explicitly now that OFN_ENABLEHOOK is set (tested - works for me).

I wonder if the UI sync could have been done somewhat different, for example by attaching a timer to Geany main window instead of the FileOpen window, and processing one gtk+ event (if any) on each timer message. "for (i = 0; i < 4 && gtk_events_pending(); i++)" is what bugs me - why 4? Will 4 it work for everyone?

> Anyway, so if I read you correctly, you say we can just remove the OPENFILENAME_SIZE_VERSION_400 references and all will be good?

Yes.

> Cant this cause problems if building e.g. on Windows 8 and trying to run on e.g. XP? (as the size would be the one from Windows 8)

No. If your API targets 8 by default, the structure may be larger, but XP will simply ignore the rest (and it'll be memset to 0). This "first member is size" approach is common to all version-dependent windows structures.

The description of GetOpenFileName also states that starting with Vsita, there is some new way to get file names, so this function is unlikely to change - but we may wish to try the new API at some point, for programs running under Vista or later.

**Comment by codebrainz** [comment]
>  "for (i = 0; i < 4 && gtk_events_pending(); i++)" is what bugs me - why 4? Will 4 it work for everyone?

Me too, that's why I left those comments :) IIRC 4 could be any number between 1 and N where N locks up the whole OS. I don't remember exactly but probably processing events on the main loop caused more events, so it just kept going, so it needed to handle a few events to update the UI but not enough to break Windows.

IMO, the best way to handle it would be to integrate the dialog loop into the GLib main loop like GTK does for the main window.

**Comment by eht16** [comment]
I guess two of the mentioned changes should be harmless and actually fix this issue:
  * remove OPENFILENAME_SIZE_VERSION_400 hacks
  * set OFN_ENABLESIZING for resizable dialogs

@zhekov since these were your ideas, do you want to open a PR? If not, I could do so as well, just don't want to steal your work :)

**Comment by zhekov** [comment]
> zhekov since these were your ideas, do you want to open a PR? If not, I could do so as well, just don't want to steal your work :)

I'm past the period where my name attached to a trivial patch made me proud. :) Please open a PR and commit them, git still makes me a bit uneasy.

There are some optimizations which can be done in win32 - for example, several members of OPENFILENAME are always initialized identically, and the check for a Windows file open dialog error is identical, but they are duplicated 5 times. Since I want to check if the UI sync can be done easier, I'll probably take care of these as well, in a future PR.
